aodntools/timeseries_products/gridded_timeseries.py

```diff
@@ -0,0 +1,262 @@
+            depth_mask[depth_min:depth_max] = [True] * depth_interval_len
+    return depth_mask
+
+def sort_attributes(attributes):
+    """
+
+    :param attributes: dictionary of global attributes
+    :return: sorted dictionary
+    """
+    return dict(sorted(attributes.items()))
+
+def sort_depths(depths, values):
+    """
+    Sort the list of depths and values
+    """
+    index =list(range(len(depths)))
+    index.sort(key=depths.__getitem__)
+    depths[:] = [depths[i] for i in index]
+    values[:] = [values[i] for i in index]
+
+    return depths, values
+
+
+
+
+## MAIN FUNCTION
+def grid_variable(file_name, VoI, depth_bins=None, max_separation=50, depth_bins_increment=10,
+                  base_path='./'):
+    """
+    Grid VoI into depth_bins.
+    :param nc: hourly aggregated dataset with VoI, DEPTH and TIME only
+    :param VoI: variable of interest (TEMP or PSAL)
+    :param depth_bins: list of depth where to interpolate. if null list is provided it will be calculated from the data
+    :param max_separation: max separation allowed for instruments
+    :param depth_bins_increment: in case no depth bins provided this is the increment for the calculated bins
+    :param base_path: path where the result file will be written
+    :return: interpolated dataset
+    """
+
+    with xr.open_dataset(file_name) as nc:
+        site_code = nc.site_code
+        ## get global attributes
+        global_attributes = nc.attrs
+
+        ## get lat/lon
+        longitude_mean = nc.LONGITUDE.mean()
+        latitude_mean = nc.LATITUDE.mean()
+
+        ## in case no depth bins provided, create depth bins to the nearest rounded depth
+        ## if provided, adjust to the min-max registered depth
+        if not depth_bins:
+            depth_bins = make_depth_bins(nc=nc, increment=depth_bins_increment)
+        else:
+            depth_bins = [float(depth) for depth in depth_bins]     # in case depth_bins provided through inline arguments
+            depth_bins = adjust_depth_bins(depth_bins, nc.DEPTH.min(), nc.DEPTH.max())
+
+        ## create empty containers
+        time_0 = pd.to_datetime('1950-01-01T00:00:00')
+        time_min = nc.TIME.values.min()
+        depth_bin_len = len(depth_bins)
+        VoI_binned = xr.DataArray(np.full((depth_bin_len, 1), np.nan), coords=[[time_0], depth_bins],
+                                  dims=['TIME', 'DEPTH'])
+        VoI_ndepths = xr.DataArray(np.full(1, 0), coords=[[time_0]], dims=['TIME'])
+
+        ## group nc by individual timestamps
+        VoI_grouped = nc.groupby('TIME')
+
+
+        for timestamp, (name, group) in enumerate(VoI_grouped):
+            time = [name]
+            n_depths = int(len(group[VoI]))
+
+            if n_depths >= 2:
+                VoI_values = list(group[VoI].values)
+                depth = list(group.DEPTH.values)
+                ## sort depths
+                depth, VoI_values = sort_depths(depth, VoI_values)
+
```

Comment:
Again - your choice of using lists instead of arrays is creating more work - you convert an array to list (```depth```) then you mutate a list in sort_depths with e a more verbose method. You are saved because list(group.DEPTH.values) copies the values but this could be simpler with just 

```sort_depths(group.DEPTH.values,VoI_values)```

Reply:
As strange as it may appear, the list approach --convert the np.array to a list and using the depth sort as a list, is a bit faster (5%) than the pure numpy approach. I've used a sample file with ~12K timestamps, which is relatively small compared with the full size file that could have ~100K timestamps.